hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java

```diff
@@ -348,16 +345,15 @@ private long acquireLock() throws UnknownHostException, TException, InterruptedE
                 LOG.warn("Interrupted while waiting for lock.", e);
               }
             }, TException.class);
-      } catch (WaitingForLockException waitingForLockException) {
-        timeout = true;
-        duration = System.currentTimeMillis() - start;
       }
-    }
-
-    // timeout and do not have lock acquired
-    if (timeout && !state.get().equals(LockState.ACQUIRED)) {
+    } catch (WaitingForLockException waitingForLockException) {
+      // timeout and do not have lock acquired
       throw new CommitFailedException("Timed out after %s ms waiting for lock on %s.%s",
-          duration, database, tableName);
+          System.currentTimeMillis() - start, database, tableName);
+    } finally {
+      if (!state.get().equals(LockState.ACQUIRED)) {
+        unlock(Optional.of(lockId));
```

Comment:
Is this correct? If the lock wasn't acquired, why would it be correct to unlock it?

Reply:
While client calling lock, hive metastore creates an exclusive lock with `WAITING` state on the hive table first. If there is no other lock on the table, then metastore changes the state to `ACQUIRED`. Otherwise, updates `hl_blockedby_ext_id` to the latest lockId. No matter whether the lock state is acquired, the lock information will stored in HIVE_LOCKS as below.

hl_lock_ext_id | hl_table | hl_lock_state | hl_lock_type | hl_last_heartbeat | hl_blockedby_ext_id
-- | -- | -- | -- | -- | --
1651514 | test_table | a | e | 1614062112285 |  
1651516 | test_table | w | e | 1614062132342 | 1651514
1651518 | test_table | w | e | 1614062152426 | 1651516

In this case, a new lock will remain in the waiting state until the latest lock is deleted by client or txn cleanup thread.

Reply:
Correct, `HiveMetaStoreClient.lock()` will enqueue a new lock request into the HMS `HIVE_LOCKS` table and if we throw an exception here upon timeout, our lock request will be stuck there in `WAITING` state (potentially blocking other subsequent requests) unless we call `unlock()` which makes sure to remove the lock from the HMS table.

Reply:
The cleanup process will remove these locks when the timeout is reached, but until that time this can block other lock requests. So it is good to clean these locks when possible

Reply:
Okay, and passing the lock ID into unlock ensures that this is only for the lock request that was created when trying to acquire. Sounds good to me. Thanks for the explanation, @ZorTsou!